src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -34,24 +34,51 @@ class ShenandoahFreeSet : public CHeapObj<mtGC> {
   ShenandoahHeap* const _heap;
   CHeapBitMap _mutator_free_bitmap;
   CHeapBitMap _collector_free_bitmap;
+  // We keep the _old_collector regions separate from the young collector regions.  This allows us to pack the old regions
+  // further to the right than the young collector regions.  This is desirable because the old collector regions are recycled
+  // even less frequently than the young survivor regions.  In generational mode, the young survivor regions are typically
+  // recycled after tenure age GC passes.
+  CHeapBitMap _old_collector_free_bitmap;
   size_t _max;
 
   // Left-most and right-most region indexes. There are no free regions outside
   // of [left-most; right-most] index intervals
   size_t _mutator_leftmost, _mutator_rightmost;
   size_t _collector_leftmost, _collector_rightmost;
+  size_t _old_collector_leftmost, _old_collector_rightmost;
 
   size_t _capacity;
+  size_t _old_capacity;
   size_t _used;
 
+  // When is_old_collector_free regions sparsely populate the lower address ranges of the heap, we search from left to
+  // right in order to consume (and remove from the is_old_collector_free range) these sparsely distributed regions.
+  // This allows us to more quickly condense the range of addresses that represent old_collector_free regions.
+  bool _old_collector_search_left_to_right = true;
+
+
   void assert_bounds() const NOT_DEBUG_RETURN;
 
-  bool is_mutator_free(size_t idx) const;
-  bool is_collector_free(size_t idx) const;
+  inline bool is_mutator_free(size_t idx) const;
+  inline bool is_collector_free(size_t idx) const;
+  inline bool is_old_collector_free(size_t idx) const;
```

Comment:
Would it be correct to assume (when we are stable between transitions in set membership) that the set of all free regions is tiled by its mutually exclusive subsets `collector_free`, `old_collector_free`, and `mutator_free` ?

Thus, when set membership is not being changed, I'd expect a free region to be in exactly one of these three subsets? If so, then (re `peek` methods below) one might be able to do membership assertion checks before and after transitions in membership (i.e. before removing from one subset and after adding to another subset). That would then avoid having to use `peek` versions of the membership querying methods.

I am trying to get a high level mental model so I can better understand the implementation and usage details. This is likely documented somewhere in `ShenandoahFreeSet`, but I am having trouble locating such documentation in the files I am looking at.

Reply:
I've added documentation to make this more clear, I hope.